src/managers/health.py

```diff
@@ -0,0 +1,63 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Manager for handling Kafka machine health."""
+
+import logging
+
+from requests.exceptions import ConnectionError, HTTPError
+
+from core.cluster import SUBSTRATES, ClusterState
+from core.workload import WorkloadBase
+from exceptions import OSDAPIError
+from literals import (
+    MSG_STATUS_ERROR,
+    MSG_STATUS_UNAVAIL,
+    MSG_STATUS_UNHEALTHY,
+    MSG_STATUS_UNKNOWN,
+    MSG_STATUS_WORKLOAD_DOWN,
+)
+from managers.api import APIManager
+
+logger = logging.getLogger(__name__)
+
+
+class HealthManager:
+    """Manager for handling Kafka machine health."""
+
+    def __init__(
+        self,
+        state: ClusterState,
+        workload: WorkloadBase,
+        substrate: SUBSTRATES,
+    ):
+        self.state = state
+        self.workload = workload
+        self.substrate = substrate
+        self.api_manager = APIManager(state, workload, substrate)
+
+    def status_ok(self) -> tuple[bool, str]:
+        """Health status"""
+        try:
+            status_data = self.api_manager.service_status()
+        except HTTPError as err:
+            if err.response.status_code == 503:
+                return False, MSG_STATUS_UNAVAIL
+        except (ConnectionError, OSDAPIError):
+            return False, MSG_STATUS_UNAVAIL
+
+        if status_data["status"]["overall"]["state"] == "green":
+            return True, ""
+        elif status_data["status"]["overall"]["state"] == "yellow":
```

Comment:
what does `yellow` or other colors refer to?

Reply:
How I wish this was documented... :-( 

Based on Opensearch, I assumed there may be a `yellow` (degraded) state. As I assume development of this API went much alongside of development of Elastic.

Now I find this: https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/server/status/legacy_status.ts#L114-L169
Let me know if you would like us to "hardcode" more of this into the dashboards. I have no idea how much we can rely on these values as I find no docs (that could assure none of this changes across versions...)

Reply:
(Regarding testing: I'm feeling a bit lost how to "squeeze out" these statuses from OSD -- since to obvious anomalies, like Opensearch being down the OSD service itself also just goes down. Thus I'm unsure how to have integration tests or even manual testing for states of various colors. Of course unit testing is already there for 'yellow', and I could extend that in case we decide to handle the full granularity.)

Alternatively we can stick to a "rule of thumb" policy as the current one:
 - green - good
 - yellow - mark it in the message but don't change unit status to blocked
 - anything else - unit is blocked, we report error
 
(I assume this may safe enough to hold across version changes and so.)

Reply:
In that case, perhaps:
- green: good 
- yellow: `WaitingStatus(waiting for service health to improve...)`
- red: BlockedStatue

Reply:
> unsure how to have integration tests or even manual testing for states of various colors

For green, it's clear. However for "red" or opensearch unreachable, I would have an integration test where, we:
1. juju ssh into all `opensearch`units
2. send a `SIGSTOP` signal to opensearch
3. ensure that all the opensearch nodes are indeed not online
4. wait for status of the dashboards to reflect the blocked state (opensearch server not reachable)
5. juju ssh into all opensearch units 
6. send a `SIGCONT` signal to opensearch
7. wait for the nodes to be online 
8. wait for status of the dashboards to be active 
9. check that units of the dashboards are online (http 200)  

For the opensearch part, you can refer to this HA test 
https://github.com/canonical/opensearch-operator/blob/main/tests/integration/ha/test_ha.py#L262-L264

Reply:
I see your point, but the above does not test the `api/status` "red" state.

The above test --from Opensearch Dashboards service pespective -- is almost equal to the existing test, where we break the relation to Opensearch.

Opensearch Dashboards responds to both situations the same way. Service goes down. There is NO `api/status`. Not in any color.

I can implement the test if you still prefer, however in terms of the healthcheck it is redundant to the already existing healthcheck. As the healthcheck --in both cases-- finds that the API is down.

However I see a valid use case that is currently not covered -- which the test above is spotting on.
In addition to the charm relation check, we should check Opensearch status (via an Opensearch REST API query). This functionality is indeed missing from the current solution. And this is to be detected by the test.

How should we check on Opensearch status? Accessing all known endpoints if they are all down?

Reply:
So to summarize the offline discussion:

 - indeed we're missing Opensearch status detection (Opensearch API status). To be added.
   - with the test above
 - another missing functionality was discovered in the meanwhile: version mismatch detecion. To be added.

Reply:
Updates: both added.